internal/pkg/agent/application/upgrade/step_unpack.go

```diff
@@ -7,52 +7,103 @@ package upgrade
 import (
 	"archive/tar"
 	"archive/zip"
+	"bytes"
 	"compress/gzip"
 	"fmt"
 	"io"
 	"io/fs"
 	"os"
+	"path"
 	"path/filepath"
 	"runtime"
 	"strings"
 
 	"github.com/hashicorp/go-multierror"
 
 	"github.com/elastic/elastic-agent/internal/pkg/agent/errors"
+	v1 "github.com/elastic/elastic-agent/pkg/api/v1"
 	"github.com/elastic/elastic-agent/pkg/core/logger"
 )
 
+type UnpackResult struct {
+	Hash string `json:"hash" yaml:"hash"`
+	// TODO add mapped path of executable
+	// agentExecutable string
+	VersionedHome string `json:"versioned-home" yaml:"versioned-home"`
+}
+
 // unpack unpacks archive correctly, skips root (symlink, config...) unpacks data/*
-func (u *Upgrader) unpack(version, archivePath, dataDir string) (string, error) {
+func (u *Upgrader) unpack(version, archivePath, dataDir string) (UnpackResult, error) {
 	// unpack must occur in directory that holds the installation directory
 	// or the extraction will be double nested
-	var hash string
+	var unpackRes UnpackResult
 	var err error
 	if runtime.GOOS == windows {
-		hash, err = unzip(u.log, archivePath, dataDir)
+		unpackRes, err = unzip(u.log, archivePath, dataDir)
 	} else {
-		hash, err = untar(u.log, version, archivePath, dataDir)
+		unpackRes, err = untar(u.log, version, archivePath, dataDir)
 	}
 
 	if err != nil {
-		u.log.Errorw("Failed to unpack upgrade artifact", "error.message", err, "version", version, "file.path", archivePath, "hash", hash)
-		return "", err
+		u.log.Errorw("Failed to unpack upgrade artifact", "error.message", err, "version", version, "file.path", archivePath, "unpackResult", unpackRes)
+		return UnpackResult{}, err
 	}
 
-	u.log.Infow("Unpacked upgrade artifact", "version", version, "file.path", archivePath, "hash", hash)
-	return hash, nil
+	u.log.Infow("Unpacked upgrade artifact", "version", version, "file.path", archivePath, "unpackResult", unpackRes)
+	return unpackRes, nil
 }
 
-func unzip(log *logger.Logger, archivePath, dataDir string) (string, error) {
+func unzip(log *logger.Logger, archivePath, dataDir string) (UnpackResult, error) {
 	var hash, rootDir string
 	r, err := zip.OpenReader(archivePath)
 	if err != nil {
-		return "", err
+		return UnpackResult{}, err
 	}
 	defer r.Close()
 
 	fileNamePrefix := strings.TrimSuffix(filepath.Base(archivePath), ".zip") + "/" // omitting `elastic-agent-{version}-{os}-{arch}/` in filename
 
+	pm := pathMapper{}
+	versionedHome := ""
+
+	// Load manifest, the use of path.Join is intentional since in .zip file paths use slash ('/') as separator
+	manifestFile, err := r.Open(path.Join(fileNamePrefix, "manifest.yaml"))
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		// we got a real error looking up for the manifest
+		return UnpackResult{}, fmt.Errorf("looking up manifest in package: %w", err)
+	}
+	if err == nil {
+		// load manifest
+		defer manifestFile.Close()
```

Comment:
should this maybe be called directly instead of with a defer? defer will wait to the end of the function, the contents seems to be read here and done. I would prefer to see this closed as soon as possible